web/src/user_groups.ts

```diff
@@ -316,6 +316,21 @@ export function get_recursive_subgroups(target_user_group: UserGroup): Set<numbe
     return subgroup_ids;
 }
 
+export function is_subgroup_of_target_group(target_group_id: number, subgroup_id: number): boolean {
+    const target_user_group = get_user_group_from_id(target_group_id);
+    const direct_subgroup_ids = new Set(target_user_group.direct_subgroup_ids);
+    if (direct_subgroup_ids.has(subgroup_id)) {
+        return true;
+    }
+
+    const recursive_subgroup_ids = get_recursive_subgroups(target_user_group);
+    if (recursive_subgroup_ids === undefined) {
+        return false;
+    }
+
+    return recursive_subgroup_ids.has(subgroup_id);
```

Comment:
Why does this check both `direct_subgroup_ids` and `get_recursive_subgroups` together? I would expect `get_recursive_subgroups` to include `direct_subgroup_ids` in what it returns.

Reply:
That helps in avoiding `get_recursive_subgroups` computation if group is direct subgroup which just involves checking if the element is in set.

Reply:
I see, so it's a performance thing? Not likely to matter. I think if we want to work on such performance details, we can use better algorithms in various places that are careful to traverse data structures only once, but I don't think it's important given the scale of groups we expect to do that at this time.